### PR TITLE
[ENG-1937] Allow SHARE to view registration data

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Move subclassed share fields to superclassed abstractprovider and ensure an access token is sent from the Registration Provider to share.

Note:

There is probably more test changes and general clean-up for this ticket then is strictly necessary, but while developing this I had a lot of difficulty with the mocking and decided to standardize the mocking to use a `responses` fixture and run celery live to ensure test were comprehensive. I have listed these changes under "Non-Functional requirements" in changes.

## Changes

Functional requirements:
- moves field location to super model and add migration
- Sends SHARE data from registration provider.

Non-Functional requirements:
- Makes all SHARE usage configurable so it's easily disabled/enabled for local development via the SHARE_ENABLED setting.
- DRYs up many SHARE functions and makes a pytest fixture that mocks SHARE is a more reliable way.
- Moves share specific tests into their own directory and removes many patches and mocks.
- `share_publish_type` is now takes advantage of the provider standardization and no longer needs to be passed specially through many functions.

## QA Notes

This effects the AbstractProvider Model and Provider API Endpoints

- Does this change require a data migration? Yes a small one
- What is the level of risk? Low
- Any permissions code touched? No

 QA can verify this via the admin app.

## Documentation

technical task, no new docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1937

### CODE REVIEW

**Comment by aaxelb** [comment]
what @brianjgeiger said
> Looks like you hit the first two parts of the ticket, but I'm not seeing the third, "when pushing registrations to SHARE, make sure the source field is used, so the registration show up in SHARE under the appropriate provider"

[how to do that is detailed in this jira comment](https://openscience.atlassian.net/browse/ENG-1937?focusedCommentId=76571)

File: website/project/tasks.py [hunk 1]
```diff
@@ -46,24 +50,22 @@ def update_collecting_metadata(node, saved_fields):
...
+        resp = send_share_json(node, data)
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError:
+            if resp.status_code >= 500:
+                _async_update_node_share.delay(node._id)
+            else:
+                send_desk_share_error(node, resp, 0)
 
-def _update_node_share(node):
```
Comment (Johnetordoff) [inline website/project/tasks.py:56]: Remove this was basically just a stylist choice

File: website/preprints/tasks.py [hunk 2]
```diff
@@ -76,22 +82,18 @@ def update_or_enqueue_on_preprint_updated(preprint_id, update_share=True, share_
 
 def update_preprint_share(preprint, old_subjects=None, share_type=None):
     if settings.SHARE_URL:
-        if not preprint.provider.access_token:
```
Comment (Johnetordoff) [inline website/preprints/tasks.py:79]: Now in the unlikely event this is set to None or False, we use the default token instead.

File: osf_tests/test_node.py [hunk 3]
```diff
@@ -3860,23 +3862,18 @@ def test_queueing_on_node_updated(self, node, user):
...
 
-        kwargs = requests.post.call_args[1]
-        graph = kwargs['json']['data']['attributes']['data']['@graph']
+        data = json.loads(responses.calls[0].request.body.decode())
+        graph = data['data']['attributes']['data']['@graph']
 
-        assert requests.post.called
-        assert kwargs['headers']['Authorization'] == 'Bearer Token'
-        assert graph[0]['uri'] == '{}{}/'.format(settings.DOMAIN, node._id)
+        assert responses.calls[0].request.headers['Authorization'] == 'Bearer mock-api-token'
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf_tests/test_node.py:3872]: Yes, that's an excellent idea, didn't see the forest through the trees!

File: conftest.py [hunk 4]
```diff
@@ -144,3 +144,26 @@ def teardown_es():
         teardown_es()
     else:
         yield
+
+@pytest.fixture(scope='function', autouse=True)
+def _mock_share_marker(request, es6_client):
+    """Clear out all indices and index templates before and after
+    tests marked with ``es``.
+    """
+    marker = request.node.get_closest_marker('mock_share')
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply conftest.py:153]: Yes, multiple fixtures is probably better. I just want to use `responses` to get away from selectively mocking functions to avoid hitting SHARE, that can be kind of wonky.

File: tests/test_registrations/test_retractions.py [hunk 5]
```diff
@@ -562,7 +550,6 @@ def test_approval_cancels_active_embargoes_on_descendant_nodes(self, mock_update
         for node in descendants:
             assert_true(node.is_retracted)
 
-        assert mock_update_share.called
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_registrations/test_retractions.py:565]: I'm am going try to phase out the method of selectively mocking functions for tests and focus on what actually gets sent in the request to SHARE (or whatever service). It's difficult to mock certain objects, especially client, I ran into that problem a lot on this ticket.

I'll try to add some additional assertions here, but I really want to avoid this level of method monkey mocking for features in the future, especially when responses can be used

File: conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaxelb) [reply conftest.py:153]: [looks like you could mock multiple responses in one fixture](https://github.com/getsentry/responses#responses-as-a-pytest-fixture), if that helps.

i'm pro-`responses`, seems much cleaner than selectively mocking functions, i agree 👍

File: tests/test_registrations/test_retractions.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaxelb) [reply tests/test_registrations/test_retractions.py:565]: yeah that makes sense -- was thinking you could use the `responses`-based mock (yielded from a `mock_share` fixture or similar) to assert the correct stuff got sent to SHARE

File: osf/models/provider.py [hunk 6]
```diff
@@ -29,6 +29,10 @@ class Meta:
         unique_together = ('_id', 'type')
         permissions = REVIEW_PERMISSIONS
 
+    PUSH_SHARE_TYPE_CHOICES = (('Preprint', 'Preprint'),
+                               ('Thesis', 'Thesis'),)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/provider.py:33]: I'd only be concerned that this would break share.osf.io UI elements.

[the remark this answers is not in the document]
Comment (aaxelb) [reply osf/models/provider.py:33]: yes, reasonable concern! no worries tho, `Registration` is a valid SHARE type

[the remark this answers is not in the document]
Comment (aaxelb) [reply osf/models/provider.py:33]: (if you're curious, [SHARE's types are listed here](https://share.osf.io/api/v2/schema/))

File: website/preprints/tasks.py [hunk 7]
```diff
@@ -1,23 +1,18 @@
...
-from website import settings, mails
-from website.util.share import GraphNode, format_contributor, format_subject
+from website import settings
+from api.share.utils import update_share
 
 logger = logging.getLogger(__name__)
 
+
 @celery_app.task(ignore_results=True, max_retries=5, default_retry_delay=60)
-def on_preprint_updated(preprint_id, update_share=True, share_type=None, old_subjects=None, saved_fields=None):
```
Comment (Johnetordoff) [inline website/preprints/tasks.py:20]: update_share was always set to true

File: website/project/tasks.py [hunk 8]
```diff
@@ -72,8 +74,8 @@ def _async_update_node_share(self, node_id):
     AbstractNode = apps.get_model('osf.AbstractNode')
     node = AbstractNode.load(node_id)
 
-    data = serialize_share_node_data(node)
-    resp = send_share_node_data(data)
+    data = serialize_share_data(node)
+    resp = send_share_json(data)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply website/project/tasks.py]: To address the instead of writing any entirely new tests. I simply removed some of  the mocking from `TestOnNodeUpdate` this caused the above error to manifest, so I simply fixed it so those tests now pass properly and cover this situation.

File: osf_tests/test_node.py [hunk 9]
```diff
@@ -2529,7 +2527,6 @@ def test_contributor_set_visibility_validation(self, node, user, auth):
                 {'user': reg_user2, 'permissions': ADMIN, 'visible': False},
             ]
         )
-        print(node.visible_contributor_ids)
```
Comment (Johnetordoff) [inline osf_tests/test_node.py:2532]: This was just a drive-by clean-up

File: osf_tests/test_node.py [hunk 10]
```diff
@@ -3801,25 +3798,7 @@ def node_in_collection(self, collection):
...
-        NodeFactory(
-            creator=node.creator,
-            parent=node,
-            title='Title1',
-        )
-        registration = RegistrationFactory(project=node)
-        registration.refresh_from_db()
-        return registration.get_nodes()[0]
-
-    def teardown_method(self, method):
```
Comment (Johnetordoff) [inline osf_tests/test_node.py:3819]: apparently this method isn't really necessary.

File: api/share/utils.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+    try:
+        resp.raise_for_status()
+    except requests.HTTPError:
+        if resp.status_code >= 500:
+            async_update_resource_share.delay(resource._id, old_subjects)
+        else:
+            send_desk_share_error(resource, resp, 0)
+
+@celery_app.task(bind=True, max_retries=4, acks_late=True)
+def async_update_resource_share(self, guid, old_subjects=None):
```
Comment (Johnetordoff) [inline api/share/utils.py:332]: Consolidating `async_update_node_share` and `async_update_prepriny_share`, might seem like a little scope creepy, but when we are asynchronously making external requests we really want to cut down on the number of moving parts and redundancies, especially because of the mocking and coverage issues present in the ticket.

File: api/share/utils.py [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        uri=subject.absolute_api_v2_url,
+    )
+    context[subject.id].attrs['parent'] = format_subject(subject.parent, context)
+    context[subject.id].attrs['central_synonym'] = format_subject(subject.bepress_subject, context)
+    return context[subject.id]
+
+
+def send_share_json(resource, data):
+    if getattr(resource, 'provider'):
+        access_token = resource.provider.access_token
```
Comment (aaxelb) [inline api/share/utils.py:99]: if the provider has null `access_token`, this'll cause a SHARE error. probably better to fall back to `settings.SHARE_API_TOKEN` in that case.

File: admin/registration_providers/views.py [hunk 13]
```diff
@@ -356,3 +358,40 @@ def post(self, request, *args, **kwargs):
...
+    def get(self, request, *args, **kwargs):
+        provider = RegistrationProvider.objects.get(id=self.kwargs['registration_provider_id'])
+
+        resp = requests.post(
+            f'{website_settings.SHARE_URL}api/v2/sources/',
+            json={
+                'data': {
+                    'type': 'Source',
+                    'attributes': {
+                        'homePage': provider.domain if provider.domain else f'{website_settings.DOMAIN}/registration/{provider._id}/',
```
Comment (aaxelb) [inline admin/registration_providers/views.py:376]: i think this is the wrong url for a registration provider
```suggestion
                        'homePage': provider.domain if provider.domain else f'{website_settings.DOMAIN}/registries/{provider._id}/',
```

File: admin/registration_providers/views.py [hunk 14]
```diff
@@ -356,3 +358,40 @@ def post(self, request, *args, **kwargs):
...
+        resp_json = resp.json()
+
+        provider.share_source = resp_json['data']['attributes']['longTitle']
+        for data in resp_json['included']:
+            if data['type'] == 'ShareUser':
+                provider.access_token = data['attributes']['token']
+
+        provider.save()
+
+        return redirect(reverse_lazy('registration_providers:share_source', kwargs={'registration_provider_id': provider.id}))
```
Comment (aaxelb) [inline admin/registration_providers/views.py:397]: the guts of this (making the request to share) duplicates [`ShareSourcePreprintProvider`](https://github.com/CenterForOpenScience/osf.io/blob/b80f2f50bb54a8c5b5259effe4d5437a74e021ef/admin/preprint_providers/views.py#L391-L435), which'll cost us in maintenance -- better to extract it to a shared function, or similar

also these safety checks from `ShareSourcePreprintProvider` (or something like them) might be nice to preserve in that shared place:
```py
        if preprint_provider.share_source or preprint_provider.access_token:
            raise ValueError('Cannot update share_source or access_token because one or the other already exists')
        if not osf_settings.SHARE_API_TOKEN or not osf_settings.SHARE_URL:
            raise ValueError('SHARE_API_TOKEN or SHARE_URL not set')
        if not preprint_provider.get_asset_url('square_color_no_transparent'):
            raise ValueError('Unable to find "square_color_no_transparent" icon for provider')
```

File: admin/providers/views.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        if not provider.get_asset_url('square_color_no_transparent'):
+            raise ValueError('Unable to find "square_color_no_transparent" icon for provider')
+
+        resp = requests.post(
+            f'{settings.SHARE_URL}api/v2/sources/',
+            json={
+                'data': {
+                    'type': 'Source',
+                    'attributes': {
+                        'homePage': provider.domain if provider.domain else f'{settings.DOMAIN}/registration/{provider._id}/',
```
Comment (aaxelb) [inline admin/providers/views.py:32]: `/registration/` here is wrong -- it'll vary by provider type: `/registries/` and `/preprints/`

ideally we could get the "home page" url from the provider model, but failing that could just build it in each view and add a second parameter to that share'd function
```py
def setup_share_source(provider, provider_home_page):
```

File: admin/providers/views.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+from osf.models import AbstractProvider
+from django.contrib.auth.mixins import PermissionRequiredMixin
+from django.views.generic import View
+from django.urls import reverse_lazy
+from django.shortcuts import redirect
+
+
+class ShareSourceProvider(PermissionRequiredMixin, View):
+    view_name = 'share_source'
+    permission_required = 'osf.change_registrationprovider'
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply admin/providers/views.py]: I moved the actual guts of the view into the abstractprovider model, so that function is shared, but admin app views separate and very short.

File: admin_tests/nodes/test_views.py [hunk 17]
```diff
@@ -277,29 +279,33 @@ def setUp(self):
...
-    @mock.patch('website.project.tasks.send_share_node_data')
-    def test_reindex_node_share(self, mock_update_share):
+    def test_reindex_node_share(self):
         count = AdminLogEntry.objects.count()
         view = NodeReindexShare()
         view = setup_log_view(view, self.request, guid=self.node._id)
-        view.delete(self.request)
+        with mock.patch('api.share.utils.settings.SHARE_ENABLED', True):
+            with mock.patch('api.share.utils.settings.SHARE_API_TOKEN', 'mock-api-token'):
+                with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
```
Comment (aaxelb) [inline admin_tests/nodes/test_views.py:288]: not important, but i just happened to learn that in python 3.1+ you can give several context managers to a single `with` instead of nesting multiple:
```py
with mock.patch('api.share.utils.settings.SHARE_ENABLED', True),
    mock.patch('api.share.utils.settings.SHARE_API_TOKEN', 'mock-api-token'),
    responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
```
...but splitting it into multiple lines like that [might only work in 3.10+](https://bugs.python.org/issue12782)? like i said, not important

**Comment by aaxelb** [comment]
(oh but also there are conflicts 😢)

**Comment by brianjgeiger** [comment]
Oh, and there are failing tests? I thought I saw all these working earlier. Maybe with the last develop merge?

File: osf/migrations/0209_auto_20200701_1658.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+# Generated by Django 1.11.28 on 2020-07-01 16:58
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0208_update_EGAP_schema'),
```
Comment (brianjgeiger) [inline osf/migrations/0209_auto_20200701_1658.py:11]: This migration shouldn't be here. It should be after 0211_auto_20200709_1320.py.

File: osf/migrations/0209_conference_auto_check_spam.py [hunk 19]
```diff
@@ -8,7 +8,7 @@
 class Migration(migrations.Migration):
 
     dependencies = [
-        ('osf', '0208_update_EGAP_schema'),
+        ('osf', '0209_auto_20200701_1658'),
```
Comment (brianjgeiger) [inline osf/migrations/0209_conference_auto_check_spam.py:11]: This migration shouldn't be in this PR.

File: osf/migrations/0209_auto_20200701_1658.py (hunk 18 above)
Comment (brianjgeiger) [reply osf/migrations/0209_auto_20200701_1658.py]: Sorry, I had master checked out. It should be after 0213_auto_20200728_1609.py.

**Comment by brianjgeiger** [comment]
Don't you still need `osf/migrations/0209_auto_20200701_1658.py`, doesn't it have the field changes for this PR?

Also, go to https://github.com/CenterForOpenScience/osf.io/pull/9413/files and do a search for `busara` - those are the binary files I'm talking about that probably shouldn't be in this PR.
